Fetch available BOM versions and show them in UI

etanshaul opened the pull request:
progress towards #1921 

- Adds service to fetch BOM versions from Maven Central (using Aether abstractions)
- Populates the BOM versions (the latest n of them) in the UI
- Hides this with a feature flag (`feature.bom`)
- If no versions can be fetched from Maven Central, then the BOM UX is hidden (with the idea that instead we will fall back to the versions from the static json metadata with no BOM support).

Does **not** have any of the following (will come in subsequent PRs):
- Adding the BOM to the pom.xml
- syncing with existing BOM
- updating the versions in the UX for each library to match what is in the selected BOM

![image](https://user-images.githubusercontent.com/1735744/37163967-1612302e-22c7-11e8-8cc5-90406b85c52a.png)

etanshaul commented:
@ivanporty @nbashirbello @patflynn ptal

ivanporty commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/GoogleCloudApiSelectorPanel.java:
```diff
@@ -99,6 +108,12 @@
+    if (ServiceManager.getService(PluginInfoService.class).shouldEnable(GctFeature.BOM)) {
+      populateBomVersions();
+    } else {
+      hideBomUI();
```
When the feature is off, and components are hidden, does the panel adjust itself properly, i.e., there is no large empty spaces?

etanshaul replied:
yes. Ive confirmed it

ivanporty commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/GoogleCloudApiSelectorPanel.java:
```diff
@@ -224,6 +249,27 @@ private void updateManagementUI() {
+        CloudApiMavenService.getInstance().getBomVersions(NUM_BOM_VERSIONS_TO_SHOW);
+
+    if (bomVersions.isEmpty()) {
+      hideBomUI();
```
So per your description, this will hide BOM versions and UI controls, and instead show individual versions and show them in library details? In this case, do we also plan to add individual versions to POM file until BOM is available, correct?

etanshaul replied:
yes. an empty return from the BOM service call will do the following:

1) hide the BOM label and dropdown (done in this PR)
2) in the client lib selector UI, show the versions from the metadata json instead of from the BOM (effectively already in place because fetching lib versions based on the BOM is not yet implemented)
3) add the dependencies to the pom.xml as if there is no BOM support at all (like we already do)

ivanporty commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/GoogleCloudApiSelectorPanel.java:
```diff
@@ -224,6 +249,27 @@ private void updateManagementUI() {
+    if (bomVersions.isEmpty()) {
+      hideBomUI();
+    } else {
+      bomVersions.forEach(bomComboBox::addItem);
```
Will this always show proper version display name, i.e. `toString()` method is stable for `Version` and no separate renderer is needed?

etanshaul replied:
yes, looks like the toString method itself is smart enough to render the version properly

patflynn commented:
what do you guys think about dropping the 'client' in the client libraries
label?


On Thu, Mar 8, 2018 at 11:58 AM Etan Shaul <notifications@github.com> wrote:

> @ivanporty <https://github.com/ivanporty> @nbashirbello
> <https://github.com/nbashirbello> @patflynn <https://github.com/patflynn>
> ptal
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-intellij/pull/1938#issuecomment-371550786>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHf5HcNm5JFT-7boGM1ycRtKzEO0COXaks5tcWMbgaJpZM4Si_xS>
> .
>

etanshaul commented:
@patflynn just to confirm. do you mean both labels? or just "Google Cloud Client LIbraries Version". I could see dropping it from both.

ivanporty commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/CloudApiMavenService.java:
```diff
@@ -0,0 +1,74 @@
+
+  private static final String GOOGLE_CLOUD_JAVA_BOM_GROUP_NAME = "com.google.cloud";
+  private static final String GOOGLE_CLOUD_JAVA_BOM_ARTIFACT_NAME = "google-cloud-bom";
+  private static final String GOOGLE_CLOUD_JAVA_BOM_VERSION_CONSTRAINT = "[0,)";
```
Interesting math notation, is this what they used in Aether? Is this required or we can fetch all versions regardless of constraint?

etanshaul replied:
yes this is Aether notation.

The service call requires a version to be passed in. So we must pass in some pattern to indicate that we want them all. I toyed with the idea of using this notation itself to limit the results instead of limiting them later client-side, but we would have to pre-suppose a version syntax (which is a bad idea I think).

etanshaul commented:
oops test is broken cause I forgot to enable the feature flag for testing. fix coming

ivanporty commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/CloudApiMavenService.java:
```diff
@@ -0,0 +1,74 @@
+              GOOGLE_CLOUD_JAVA_BOM_VERSION_CONSTRAINT,
+              ArtifactKind.ARTIFACT);
+
+      availableVersions.sort(Comparator.reverseOrder());
```
Will this always work, i.e. `Version` implements `Comparable`?

etanshaul replied:
yes, version implements comparable

ivanporty commented:
Agree, library sounds clear enough without "client"

patflynn commented:
Good point.  Right now it's a bit weird that it's inconsistent between the
two.
maybe something like:
window title: "Add Google Cloud Libraries"
1st label: 'Add library to module'
2nd label: 'library version'

I think I'd also be OK with prefacing everything with 'Google Cloud'.


On Thu, Mar 8, 2018 at 12:10 PM Etan Shaul <notifications@github.com> wrote:

> @patflynn <https://github.com/patflynn> just to confirm. do you mean both
> labels? or just "Google Cloud Client LIbraries Version". I could see
> dropping it from both.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-intellij/pull/1938#issuecomment-371554806>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHf5HdtY2q9EusweIVQbTZHNnzK3kYm_ks5tcWXogaJpZM4Si_xS>
> .
>

etanshaul commented:
@patflynn 

I like the suggestions. Regarding prefacing everything with Google Cloud - the 2nd label imo must have that preface otherwise its a bit weird. But I'm also OK prefacing them all with it.

ivanporty commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/CloudApiMavenService.java:
```diff
@@ -0,0 +1,74 @@
+   * Returns the available Google Cloud Java client library BOM versions from Maven Central.
+   *
+   * @param limit the number of results to the n latest BOM versions
+   * @return returns the {@link Version versions} of the BOMs
```
Probably worth mentioning this is reversely sorted list.

etanshaul replied:
what do you think about moving the sorting outside of this method. seems like sorting is the responsibility of the client that is displaying them, not the service itself.

then we can even unit test this here in this pr

ivanporty replied:
Sounds like a great idea, it does make two things here and sorting is needed for UI only.

ivanporty commented on google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/apis/GoogleCloudApiSelectorPanelTest.java:
```diff
@@ -235,6 +244,37 @@ public void getPanel_withMultipleLibraries_multipleSelected_hasCheckboxesAndPopu
+    assertThat(panel.getBomComboBox().isVisible()).isTrue();
+    assertThat(bomComboBox.getItemCount()).isEqualTo(3);
+
+    for (int i = 0; i < bomComboBox.getItemCount(); i++) {
```
Nit: remembering Java readability from Nate :) it could be cleaner to move this to a separate test case and test against explicit list of values rather than a loop.

etanshaul replied:
sounds good

etanshaul replied:
the ghost of Nate haunts us :)

etanshaul replied:
I removed the loop. but didn't see much value in moving to a separate test case..lmk if you disagree

ivanporty replied:
LGTM

etanshaul commented on google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/apis/GoogleCloudApiSelectorPanelTest.java:
```diff
@@ -574,4 +622,26 @@ private static void assertDetailsShownForLibrary(
+
+    @Override
+    public int compareTo(@NotNull Version o) {
+      return 0;
```
the sorting is done inside of the BOM service method. in these tests we are mocking out that entire method (so no sorting is done).

ivanporty replied:
cool, got it

etanshaul commented:
wording updates:
![image](https://user-images.githubusercontent.com/1735744/37166119-2ed9385a-22cc-11e8-9af0-f61980e91347.png)

ivanporty commented:
We have two places with plural "Libraries" and one with "Add .. Library to Module", is this intentional?

etanshaul commented:
Yes that is intentional:

The plural case is because that BOM version applies to ALL the client libraries. Its an overarching version that can be thought of loosely as the version of the client libraries.

ivanporty commented:
Sounds good. I just thought a user might be able to add multiple versions to module, and title says "Libraries", but it's pretty clear anyway.

etanshaul commented:
side note: I put a `todo` in the code to fetch the bom version asynchronously.

in my testing, however, its instantaneous enough to not be noticeable in the UI.

etanshaul commented:
yes - that is a bit confusing. its the difference between the BOM version and the individual library version.

If you have any ideas to make it even clearer, feel free

etanshaul commented:
@ivanporty ptal at 98035d2:

- Moved sorting out to the caller of the service
- Moved limited out to the caller (using the same reasoning as for the first point)
- Added test coverage for BOM version limiting

etanshaul commented:
updated to test reverse order sorting too

etanshaul commented:
I think I addressed all comments so far. lmk if I missed anything
